server/src/rbac/user.rs

```diff
@@ -0,0 +1,87 @@
+use std::collections::HashMap;
+
+use argon2::{
+    password_hash::{rand_core::OsRng, PasswordHasher, SaltString},
+    Argon2, PasswordHash, PasswordVerifier,
+};
+
+use rand::distributions::{Alphanumeric, DistString};
+
+use crate::option::CONFIG;
+
+#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+pub struct User {
+    pub username: String,
+    pub password: String,
+    // fill this
+    pub role: Vec<()>,
+}
+
+impl User {
+    // create a new User and return self with password generated for said user.
+    pub fn create_new(username: String) -> (Self, String) {
+        let PassCode { password, hashcode } = Self::gen_new_password();
+        (
+            Self {
+                username,
+                password: hashcode,
+                role: Vec::new(),
+            },
+            password,
+        )
+    }
+
+    // Verification works because the PasswordHash is in PHC format
+    // $<id>[$v=<version>][$<param>=<value>(,<param>=<value>)*][$<salt>[$<hash>]]
+    // ref https://github.com/P-H-C/phc-string-format/blob/master/phc-sf-spec.md#specification
+    pub fn verify(&self, password: &str) -> bool {
+        let parsed_hash = PasswordHash::new(&self.password).unwrap();
+        Argon2::default()
+            .verify_password(password.as_bytes(), &parsed_hash)
+            .is_ok()
+    }
+
+    // gen new password
+    pub fn gen_new_password() -> PassCode {
+        let password = Alphanumeric.sample_string(&mut rand::thread_rng(), 16);
+        let hashcode = gen_hash(&password);
+        PassCode { password, hashcode }
```

Comment:
if the function is called `gen_new_password` only generate the password  here and create hash in separate step.